app-jee-jsp/src/main/java/org/keycloak/quickstart/appjee/ServiceLocator.java

```diff
@@ -23,12 +23,13 @@ public static URL getServiceUrl(HttpServletRequest req) {
                 return new URL(uri);
             }
 
-            String host = new URL(req.getRequestURL().toString()).getHost();
-            if (host == null) {
-                host = "localhost";
-            }
+            URL requestUrl = new URL(req.getRequestURL().toString());
+
+            String host = requestUrl.getHost();
+            String schema = requestUrl.getProtocol();
+            String port = requestUrl.getPort() != -1 ? String.valueOf(requestUrl.getPort()) : "8080";
```

Comment:
This doesn't seem correct to me. If getPort() returns -1 it means the port is not specified, which means it's using default ports (80 or 443). Would it not be better to just take req.getRequestURL().toString() and replace the context path with '/service'?